justfile

```diff
@@ -0,0 +1,156 @@
+
+# Make sure that some env variables are set for all jobs.
+#FIXME: Consider using .env for this, but only if used also on CI
+export RUST_WORKSPACE := env_var_or_default("RUST_WORKSPACE", "discovery_engine_core")
+export DART_WORKSPACE := env_var_or_default("DART_WORKSPACE", "discovery_engine")
+export CARGO_INSTALL_ROOT := env_var_or_default("CARGO_INSTALL_ROOT", "cargo-installs")
+
+# Run just --list
+default:
+    @{{just_executable()}} --list
+
+# Get/update dart deps
+dart_deps:
+    cd "$DART_WORKSPACE"; \
+    dart pub get
+
+# Fetch rust dependencies
+rust_deps:
+    cd "$RUST_WORKSPACE"; \
+    cargo fetch --locked
+
+# Installs async-bindgen cli tool
+install_async_bindgen *args:
+    cargo install \
+        --git https://github.com/xaynetwork/xayn_async_bindgen.git \
+        {{args}} \
+        async-bindgen-gen-dart \
+        "$@" ;
+
+# Get/Update/Fetch/Install all dependencies
+deps: dart_deps rust_deps install_async_bindgen
+
+# Format dart (checks only on CI)
+dart_fmt:
+    cd "$DART_WORKSPACE"; \
+    dart format {{ if env_var_or_default("CI", "false") == "true" { "--output=none --set-exit-if-changed" } else { "" } }} .
+
+# Format rust (checks only on CI)
+rust_fmt:
+    cd "$RUST_WORKSPACE"; \
+    cargo +nightly fmt --all -- {{ if env_var_or_default("CI", "false") == "true" { "--check" } else { "" } }};\
+    cargo sort --grouped --workspace {{ if env_var_or_default("CI", "false") == "true" { "--check" } else { "" } }}
+
+# Format all code (checks only on CI)
+fmt: rust_fmt dart_fmt
+
+# Check dart code, fails on info on CI
+dart_check: dart_build
+    cd "$DART_WORKSPACE"; \
+    dart analyze {{ if env_var_or_default("CI", "false") == "true" { "--fatal-infos" } else { "" } }}
+
+_codegen_order_workaround:
+    cd "$RUST_WORKSPACE"; \
+    cargo check --quiet 2>/dev/null || :
+
+# Check rust code, fails on warnings on CI
+rust_check: _codegen_order_workaround
+    cd "$RUST_WORKSPACE"; \
+    cargo clippy --all-targets --locked #TODO DENY WARNINGS ON CI
+
+# Check rust and dart code, fails if there are any issues on CI
+check: rust_check dart_check
+
+# Check if dart documentation can be build without issues
+dart_check_doc:
+    cd "$DART_WORKSPACE"; \
+    dart pub global run dartdoc:dartdoc --no-generate-docs --no-quiet
+
+# Check if dart documentation can be build without issues
+rust_check_doc: _codegen_order_workaround
+    cd "$RUST_WORKSPACE"; \
+    cargo doc --all-features --no-deps --document-private-items --locked
+
+# Check if documentation can be build without issues
+check_doc: dart_check_doc rust_check_doc
```

Comment:
How is `check_doc` different from just generating the documentation and seeing if it fails (it looks like it's doing that). In that case it could just be called `doc`, right?

In other words, would I run this locally to generate the docs? If so, `check_doc` would be an odd name.

Reply:
It's not quite the same, dart-doc as called here doesn't generate documentation (but it will tell you it did :shrug: ).

Similar for rust we omit documentation for dependencies, but in my experience having documentation for dependencies appearing in the public interface is quite important.

I added a separate `doc` step instead.